File: duolingo_rum/Services/AchievementService.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using duolingo_rum.Models;
using Microsoft.EntityFrameworkCore;

namespace duolingo_rum.Services
{
    public class AchievementService : IDisposable
    {
        private readonly _43pRumiantsefContext _context;
        private bool _disposed = false;
        public AchievementService(_43pRumiantsefContext context)
        {
            _context = context;
        }

        public AchievementService()
        {
            _context = new _43pRumiantsefContext();
        }

        /// <summary>
        /// Проверяем все достижения после урока и выдаём незаработанные.
        /// Возвращает список НОВЫХ достижений (для показа попапа).
        /// </summary>
        public async Task<List<Achievement>> CheckAndAwardAchievements(Guid userId)
        {
            var newlyEarned = new List<Achievement>();

            try
            {
                // Загружаем юзера со статистикой
                var user = await _context.Users.FindAsync(userId);
                if (user == null) return newlyEarned;

                // Уже полученные достижения
                var alreadyEarned = await _context.UserAchievements
                    .Where(ua => ua.UserId == userId)
                    .Select(ua => ua.AchievementId)
                    .ToListAsync();

                // Все достижения из БД
                var allAchievements = await _context.Achievements.ToListAsync();

                // Считаем статистику пользователя
                var wordsLearned = await _context.WordProgresses
                    .CountAsync(wp => wp.UserId == userId);

                var totalSessions = await _context.LearningSessions
                    .CountAsync(s => s.UserId == userId);

                var totalCorrect = await _context.LearningSessions
                    .Where(s => s.UserId == userId)
                    .SumAsync(s => s.CorrectAnswers ?? 0);

                var perfectSessions = await _context.LearningSessions
                    .Where(s => s.UserId == userId && s.FinishedAt != null
                                && s.WrongAnswers == 0 && s.WordsStudied > 0)
                    .CountAsync();

                Debug.WriteLine($"CheckAchievements: words={wordsLearned}, sessions={totalSessions}, correct={totalCorrect}, perfect={perfectSessions}, streak={user.StreakDays}, xp={user.TotalXp}");

                foreach (var achievement in allAchievements)
                {
                    // Пропускаем уже полученные
                    if (alreadyEarned.Contains(achievement.Id))
                        continue;

                    bool conditionMet = achievement.ConditionType switch
                    {
                        "words_learned" => wordsLearned >= (achievement.ConditionValue ?? 0),
                        "streak_days" => (user.StreakDays ?? 0) >= (achievement.ConditionValue ?? 0),
                        "total_xp" => (user.TotalXp ?? 0) >= (achievement.ConditionValue ?? 0),
                        "sessions_count" => totalSessions >= (achievement.ConditionValue ?? 0),
                        "correct_answers" => totalCorrect >= (achievement.ConditionValue ?? 0),
                        "perfect_sessions" => perfectSessions >= (achievement.ConditionValue ?? 0),
                        _ => false
                    };

                    if (!conditionMet) continue;

                    // Выдаём достижение
                    _context.UserAchievements.Add(new UserAchievement
                    {
                        UserId = userId,
                        AchievementId = achievement.Id,
                        EarnedAt = DateTime.UtcNow
                    });

                    // Начисляем XP за достижение
                    if ((achievement.XpReward ?? 0) > 0)
                    {
                        user.TotalXp = (user.TotalXp ?? 0) + achievement.XpReward!.Value;
                        user.UpdatedAt = DateTime.UtcNow;
                    }

                    newlyEarned.Add(achievement);
                    Debug.WriteLine($"Achievement earned: {achievement.Name}");
                }

                if (newlyEarned.Count > 0)
                    await _context.SaveChangesAsync();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"CheckAndAwardAchievements ERROR: {ex.Message}");
            }

            return newlyEarned;
        }

        /// <summary>
        /// Все достижения с флагом — получено или нет.
        /// </summary>
        public async Task<List<(Achievement Achievement, bool IsEarned, DateTime? EarnedAt)>> GetAllAchievementsWithStatus(Guid userId)
        {
            try
            {
                var all = await _context.Achievements.ToListAsync();

                var earned = await _context.UserAchievements
                    .Where(ua => ua.UserId == userId)
                    .ToListAsync();

                return all.Select(a =>
                {
                    var ua = earned.FirstOrDefault(e => e.AchievementId == a.Id);
                    return (a, ua != null, ua?.EarnedAt);
                }).ToList();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetAllAchievementsWithStatus ERROR: {ex.Message}");
                return new List<(Achievement, bool, DateTime?)>();
            }
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                _context?.Dispose();
                _disposed = true;
            }
        }
    }
}

File: android-app/app/src/main/java/com/linguaai/app/services/AchievementService.kt
package com.linguaai.app.services

import android.util.Log
import com.linguaai.app.models.Achievement
import com.linguaai.app.models.UserAchievement
import com.linguaai.app.models.WordProgress
import io.github.jan.supabase.postgrest.postgrest
import java.time.Instant

class AchievementService {

    private val client = SupabaseClient.client
    private val tag = "AchievementService"

    data class AchievementWithStatus(
        val achievement: Achievement,
        val isEarned: Boolean,
        val earnedAt: String?
    )

    suspend fun checkAndAwardAchievements(userId: String): List<Achievement> {
        val newlyEarned = mutableListOf<Achievement>()

        try {
            val user = client.postgrest["users"]
                .select { filter { eq("id", userId) } }
                .decodeList<com.linguaai.app.models.User>()
                .firstOrNull() ?: return emptyList()

            val alreadyEarned = client.postgrest["user_achievements"]
                .select { filter { eq("user_id", userId) } }
                .decodeList<UserAchievement>()
                .map { it.achievementId }

            val allAchievements = client.postgrest["achievements"]
                .select()
                .decodeList<Achievement>()

            val wordsLearned = client.postgrest["word_progresses"]
                .select { filter { eq("user_id", userId) } }
                .decodeList<WordProgress>()
                .size

            val sessions = client.postgrest["learning_sessions"]
                .select { filter { eq("user_id", userId) } }
                .decodeList<com.linguaai.app.models.LearningSession>()

            val totalSessions = sessions.size
            val totalCorrect = sessions.sumOf { it.correctAnswers ?: 0 }
            val perfectSessions = sessions.count {
                it.finishedAt != null && (it.wrongAnswers ?: 0) == 0 && (it.wordsStudied ?: 0) > 0
            }

            for (achievement in allAchievements) {
                if (achievement.id in alreadyEarned) continue

                val conditionMet = when (achievement.conditionType) {
                    "words_learned" -> wordsLearned >= (achievement.conditionValue ?: 0)
                    "streak_days" -> (user.streakDays ?: 0) >= (achievement.conditionValue ?: 0)
                    "total_xp" -> (user.totalXp ?: 0) >= (achievement.conditionValue ?: 0)
                    "sessions_count" -> totalSessions >= (achievement.conditionValue ?: 0)
                    "correct_answers" -> totalCorrect >= (achievement.conditionValue ?: 0)
                    "perfect_sessions" -> perfectSessions >= (achievement.conditionValue ?: 0)
                    else -> false
                }

                if (!conditionMet) continue

                client.postgrest["user_achievements"].insert(
                    UserAchievement(
                        userId = userId,
                        achievementId = achievement.id,
                        earnedAt = Instant.now().toString()
                    )
                )

                if ((achievement.xpReward ?: 0) > 0) {
                    client.postgrest["users"].update(
                        mapOf(
                            "total_xp" to ((user.totalXp ?: 0) + (achievement.xpReward ?: 0)),
                            "updated_at" to Instant.now().toString()
                        )
                    ) {
                        filter { eq("id", userId) }
                    }
                }

                newlyEarned.add(achievement)
                Log.d(tag, "Achievement earned: ${achievement.name}")
            }
        } catch (e: Exception) {
            Log.e(tag, "checkAndAwardAchievements ERROR: ${e.message}")
        }

        return newlyEarned
    }

    suspend fun getAllAchievementsWithStatus(userId: String): List<AchievementWithStatus> {
        return try {
            val all = client.postgrest["achievements"]
                .select()
                .decodeList<Achievement>()

            val earned = client.postgrest["user_achievements"]
                .select { filter { eq("user_id", userId) } }
                .decodeList<UserAchievement>()

            val earnedMap = earned.associateBy { it.achievementId }

            all.map { achievement ->
                val ua = earnedMap[achievement.id]
                AchievementWithStatus(
                    achievement = achievement,
                    isEarned = ua != null,
                    earnedAt = ua?.earnedAt
                )
            }
        } catch (e: Exception) {
            Log.e(tag, "getAllAchievementsWithStatus ERROR: ${e.message}")
            emptyList()
        }
    }
}
